src/Build/BackEnd/Components/SdkResolution/MainNodeSdkResolverService.cs

```diff
@@ -0,0 +1,240 @@
+    /// must have access to an <see cref="IBuildComponentHost"/> and call <see cref="IBuildComponentHost.GetComponent"/> and pass <see cref="BuildComponentType.SdkResolverService"/>.
+    /// </summary>
+    internal sealed class MainNodeSdkResolverService : HostedSdkResolverServiceBase
+    {
+        /// <summary>
+        /// Stores the cache in a set of concurrent dictionaries.  The main dictionary is by build submission ID and the inner dictionary contains a case-insensitive SDK name and the cached <see cref="SdkResult"/>.
+        /// </summary>
+        private readonly ConcurrentDictionary<int, ConcurrentDictionary<string, SdkResult>> _cache = new ConcurrentDictionary<int, ConcurrentDictionary<string, SdkResult>>();
+
+        /// <summary>
+        /// An object used for locking in this class instance.
+        /// </summary>
+        private readonly object _lockObject = new object();
+
+        /// <summary>
+        /// A <see cref="Task"/> running in the background which handles requests from remote nodes.
+        /// </summary>
+        private Task _requestHandler;
+
+        /// <summary>
+        /// An event which is signaled when a request is received from a remote host.
+        /// </summary>
+        private ManualResetEvent _requestReceivedEvent;
+
+        /// <summary>
+        /// A list of requests from remote hosts which need to be processed.
+        /// </summary>
+        private ConcurrentQueue<SdkResolverRequest> _requests;
+
+        /// <summary>
+        /// A factory which is registered to create an instance of this class.
+        /// </summary>
+        public static IBuildComponent CreateComponent(BuildComponentType type)
+        {
+            return new MainNodeSdkResolverService();
+        }
+
+        /// <inheritdoc cref="ISdkResolverService.ClearCache"/>
+        public override void ClearCache(int submissionId)
+        {
+            ConcurrentDictionary<string, SdkResult> entry;
+
+            _cache.TryRemove(submissionId, out entry);
+        }
+
+        /// <inheritdoc cref="INodePacketHandler.PacketReceived"/>
+        public override void PacketReceived(int node, INodePacket packet)
+        {
+            switch (packet.Type)
+            {
+                case NodePacketType.ResolveSdkRequest:
+                    HandleRequest(node, packet as SdkResolverRequest);
+                    break;
+            }
+        }
+
+        /// <inheritdoc cref="ISdkResolverService.ResolveSdk"/>
+        public override string ResolveSdk(int submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, string solutionPath, string projectPath)
+        {
+            SdkResult result = GetSdkResult(submissionId, sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath);
+
+            return result?.Path;
+        }
+
+        /// <summary>
+        /// Resolves the specified SDK.
+        /// </summary>
```

Comment:
Maybe add a remark here about locking? This does quite a bit of implicit thread synchronization where I would probably expect to see locks. A comment about that could help make it more obvious when locks are taken via the concurrent dictionaries.